fairscale/optim/oss.py

```diff
@@ -235,14 +235,26 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
         # Sync oss param_groups attributes in case they've been updated by a scheduler.
         OSS._sync_param_groups(self.param_groups, self.optim.param_groups)
 
-        # Run the optimizer step on this shard only:
+        rank_updated = False
+        last_work_handle = None
+
         if closure is not None:
+            # Run the optimizer step on this shard only, cannot be interleaved with comms
             loss = self.optim.step(closure=closure, **kwargs)  # type: ignore
-        else:
-            loss = self.optim.step(**kwargs)
+            rank_updated = True
+
+        # Interleave optimizer steps and broadcasts, overlap comms and compute
+        for device in self.buckets.keys():
+            for src_rank, bucket in enumerate(self.buckets[device]):
+                if src_rank == self.rank and not rank_updated:
+                    loss = self.optim.step(**kwargs)
```

Comment:
I am confused. This only needs to be called once to step all params, right? why is it in this loop?

Reply:
it will only be called once (hence the `rank_updated` variable), the idea is to fire coms and compute calls in parallel instead of in a sequence

Reply:
ah wait, I'm mistaken indeed.. let me put it back to draft